### ISSUE TITLE
Serious CPU (GC) activity after phantom upgrade

### ISSUE THREAD

**Issue by daveinglis** [outworkers/phantom#588]
Hi @alexflav23 

I have finished our upgrade to phantom 1.29 (from 1.5) and it has passed all our testing so far (mostly locally against a single cassandra node).  However once we deployed it to our staging environment and our devs started using it at a much higher scale (the staging env uses a multi-node cassandra cluster), we started experiencing timeouts and just general slowness on any DB operation.  Upon investigation we discovered our module which uses phantom had very high CPU utilization and the jvm heap memory usage.

![screen shot 2016-11-24 at 9 28 41 am](https://cloud.githubusercontent.com/assets/5224890/20601801/6b69872c-b228-11e6-8023-a1e88690e4af.png)

You can see the difference once our update happened yesterday in the above graph.

So my question is, has there been any performance analysis done on phantom?  I was going to try and dig further with some JVM profiling tools but I thought I would ask before encase this is something that is being worked on.

**Comment by alexflav23** [comment]
Hi @daveinglis,

This does sound very serious indeed, we do run performance testing but definitely not as in depth as we could. The reason for that is because the Datastax Java Driver does it's own performance cycle and most of the critical IO is handled there.

Phantom sugars the execution model there with promises that latch on to completed Guava futures that exist underneath. Now I wouldn't call it impossible for locks and blocks to exist within phantom itself, though if I were to bet I'd say the issue goes deeper.

Sounds like we need to take a closer look, and we would certainly welcome any help or benchmark help you would expect to see.

Regards,
Flavian

**Comment by daveinglis** [comment]
@alexflav23, after some investigation using yourkit profiler, I am seeing some high GC activity caused by com.outworkers.diesel.engine.query.AbstractQuery with its query String that its building.  Does this look like a plausible cause to you?  I'm still investigating some other areas that have changes when comparing some profiler snapshot between and old (phantom 1.5.0) and new (1.29.4) but this class seems to be a culprit according to yourkit.

**Comment by alexflav23** [comment]
Hi @daveinglis

That would really surprise me, since that class hasn't actually changed in a very very long time and 99% of it does this: `"a" + "b" = "ab"`, it's just string concat, so it's not sounding very plausible from a distance, but who knows.`

I'm going to have a look myself in more detail, I'd be really surprised if strong concatenation would be the culprit of the application.

Regards.

**Comment by alexflav23** [comment]
Hi @daveinglis,

Have you had anymore time or made any more discoveries here? I would hate string addition to be the ultimate response as that would be bad for many reasons, but again I would imagine this is not the real culprit in question.

As we are super caught up in 2.0.0 work, further investigation on our part will have to wait one or two more weeks, but we will definitely follow up to make sure no auto-magical code was introduced to have drasttic impact on performance.

Regards,
Flavian

**Comment by daveinglis** [comment]
@alexflav23, I don't have anymore details yet, I have been putting together some specific performance tests for our platform which I can then use to better profile things in isolation.  Its also important to note that my analysis seemed to point towards just a difference in the amount of garbage collection being done (rather large difference) and not the actually string building itself (but a cause of the GC).  There was one other area that was bubbling to the top alongside the string building and that was some json serialization/deserialization that we do as well using json4s+jackson (again more related to the effect on GC).  My guess for this change was because this dependency got updated at the same time. Again, once I have better/specific tests I can move these dependencies forward on their own and measure their effect.  

Anyway, in summary I am still working on doing better analysis, it's just taking a bit longer than I hoped, so waiting a few weeks for any analysis you can do is not a problem, and is much appreciated.

Cheers
Dave

**Comment by alexflav23** [comment]
Hi @daveinglis,

I can now suggest that you try one of the more recent versions of phantom, 2.9.x respectively. We have implemented huge performance critical features that will massively improve things on a linear basis at client level.

In the interim, I will close this ticket because we have no further details to help us pursue this further and it's been quite a while, but do feel free to re-open this should there be more detail.

Regards,
Flavian
